### PR TITLE
gh-123958: move assert optimization from codegen to ast_opt

### PR DESCRIPTION
**Pull request by iritkatriel**
* Issue: gh-123958

### CODE REVIEW

File: Python/codegen.c [hunk 1]
```diff
@@ -2792,20 +2792,7 @@ codegen_from_import(compiler *c, stmt_ty s)
...
-         PyTuple_Size(s->v.Assert.test->v.Constant.value) > 0))
-    {
-        RETURN_IF_ERROR(
-            _PyCompile_Warn(c, LOC(s), "assertion is always true, "
-                                       "perhaps remove parentheses?"));
-    }
-    if (OPTIMIZATION_LEVEL(c)) {
-        return SUCCESS;
-    }
+    assert(!OPTIMIZATION_LEVEL(c));
```
Comment (iritkatriel) [inline Python/codegen.c:2795]: Left this here to convince ourselves of correctness. I'll remove it before committing.

File: Python/ast_opt.c [hunk 2]
```diff
@@ -1005,6 +1012,29 @@ astfold_stmt(stmt_ty node_, PyArena *ctx_, _PyASTOptimizeState *state)
...
+            int ret = _PyErr_EmitSyntaxWarning(msg, state->filename,
+                                               node_->lineno, node_->col_offset + 1,
+                                               node_->end_lineno, node_->end_col_offset + 1);
+            Py_DECREF(msg);
+            if (ret < 0) {
+                return 0;
+            }
+        }
+        if (state->optimize) {
+            make_pass(node_);
```
Comment (Eclips4) [inline Python/ast_opt.c:1036]: Technically speaking, this is not the removal of the `assert` statement; it's replacement of `assert` with `pass` statement. Is it possible to completely remove the `assert` node? Or will it break some existing code?

File: Lib/test/test_builtin.py [hunk 3]
```diff
@@ -491,6 +485,29 @@ async def arange(n):
...
+        code_samples = [
+            '''assert not await asyncio.sleep(0)''',
+            '''assert [x async for x in arange(1)]''',
+            '''assert {x async for x in arange(1)}''',
+            '''assert {x: x async for x in arange(1)}''',
+        ]
+        for mode, code_sample, optimize in product(modes, code_samples, optimizations):
+            with self.subTest(mode=mode, code_sample=code_sample, optimize=optimize):
+                source = dedent(code_sample)
+                if (optimize):
```
Comment (JelleZijlstra) [inline Lib/test/test_builtin.py]:
```suggestion
                if optimize:
```

File: Lib/test/test_builtin.py [hunk 4]
```diff
@@ -491,6 +485,29 @@ async def arange(n):
...
+                    co = compile(source, '?', mode, optimize=optimize)
+                    co_expected = compile('pass',  '?', mode, optimize=optimize)
```
Comment (JelleZijlstra) [inline Lib/test/test_builtin.py]:
```suggestion
                    co_expected = compile('pass', '?', mode, optimize=optimize)
```

File: Lib/test/test_builtin.py [hunk 5]
```diff
@@ -491,6 +485,29 @@ async def arange(n):
...
+                    self.assertEqual(list(co.co_lines()), list(co_expected.co_lines()))
```
Comment (JelleZijlstra) [inline Lib/test/test_builtin.py:504]: This looks like we no longer raise an error in optimized mode if there is an invalid await. That seems wrong and doesn't align with what we decided on in #121637.

Comment (iritkatriel) [reply Lib/test/test_builtin.py:504]: If we want to stick with this then we probably can't optimise the assertions in ast_opt.

File: Python/ast_opt.c (hunk 2 above)
Comment (iritkatriel) [reply Python/ast_opt.c:1036]: Pass becomes a NOP, which gets removed later. This is how we get rid of code. However, I did forget to wipe out the location information (a NOP is not removed if it's the only instruction from its line in the source code).

**Pull request closed without merging**